core/src/main/java/org/apache/iceberg/rest/RESTSessionCatalog.java

```diff
@@ -0,0 +1,626 @@
+  }
+
+  RESTSessionCatalog(Function<Map<String, String>, RESTClient> clientBuilder) {
+    this.clientBuilder = clientBuilder;
+  }
+
+  private void scheduleTokenRefresh(AuthSession session, int expiresInSeconds) {
+    // TODO: Add token refresh here
+  }
+
+  @Override
+  public void initialize(String name, Map<String, String> props) {
+    Preconditions.checkArgument(props != null, "Invalid configuration: null");
+
+    String initToken = props.get(RESTCatalogProperties.TOKEN);
+
+    // fetch auth and config to complete initialization
+    ConfigResponse config;
+    OAuthTokenResponse authResponse;
+    try (RESTClient initClient = clientBuilder.apply(props)) {
+      Map<String, String> initHeaders = merge(configHeaders(props), authHeaders(initToken));
+      String credential = props.get(RESTCatalogProperties.CREDENTIAL);
+      if (credential != null && !credential.isEmpty()) {
+        String scope = props.getOrDefault(RESTCatalogProperties.SCOPE, RESTCatalogProperties.CATALOG_SCOPE);
+        authResponse = fetchClientCredentials(initClient, initHeaders, credential, scope);
+        config = fetchConfig(initClient, merge(initHeaders, authHeaders(authResponse.token())));
+      } else {
+        authResponse = null;
+        config = fetchConfig(initClient, initHeaders);
+      }
+    } catch (IOException e) {
+      throw new UncheckedIOException("Failed to close HTTP client", e);
+    }
+
+    // build the final configuration and set up the catalog's auth
+    Map<String, String> mergedProps = config.merge(props);
+    if (authResponse != null) {
+      this.catalogAuth = new AuthSession(
+          configHeaders(mergedProps), authResponse.token(), authResponse.issuedTokenType());
+      if (authResponse.expiresInSeconds() != null) {
+        scheduleTokenRefresh(catalogAuth, authResponse.expiresInSeconds());
+      }
+    } else if (initToken != null) {
+      this.catalogAuth = new AuthSession(configHeaders(mergedProps), initToken, null);
+      // TODO: add expiration/refresh from a config property
+    } else {
+      this.catalogAuth = null;
+    }
+
+    this.client = clientBuilder.apply(mergedProps);
+    this.paths = ResourcePaths.forCatalogProperties(mergedProps);
+
+    String ioImpl = mergedProps.get(CatalogProperties.FILE_IO_IMPL);
+    this.io = CatalogUtil.loadFileIO(ioImpl != null ? ioImpl : ResolvingFileIO.class.getName(), mergedProps, conf);
+
+    super.initialize(name, mergedProps);
+  }
+
+  Supplier<Map<String, String>> headers(SessionContext context) {
+    AuthSession session = sessions.get(context.sessionId(), id -> {
+      Map<String, String> headers = catalogAuth.headers();
+
+      OAuthTokenResponse response;
+      if (context.credential() != null) {
+        response = fetchClientCredentials(
+            client, headers, context.credential(), RESTCatalogProperties.CATALOG_SCOPE);
+      } else if (context.properties().containsKey(RESTCatalogProperties.TOKEN)) {
+        response = exchangeToken(
```

Comment:
In addition to using the catalog's token in the `Authorization` header, this token exchange also uses the catalog token as the actor token.